### PR TITLE
Check Integrity year check added

### PR DESCRIPTION
**Pull request by grimes2**
#1897 Added last four nonpunctuation characters should be numerals in Check Integrity year

- [x] Change in CHANGELOG.md described
- [x] Tests created for changes
- [x] Manually tested changed features in running JabRef

### CODE REVIEW

**Comment by oscargus** [comment]
Thanks! The tests fails because of missing translation strings.

If you run the logic/i10/LocalizationConsistencyTest.java you will get a list of strings which are not in the translation files. Add those to the English one and run the script that the output tells you to. Then, all languages are updated and the test should pass (you may have to refresh the files in your IDE in case it fails again). Also, see: https://github.com/JabRef/jabref/wiki/Code-Howtos#using-localization-correctly

If you can add a test to IntegrityCheckTest (or whatever it is called) that really tests for this new feature it would be great.

**Comment by grimes2** [comment]
I hope, it is correct now. Please set label "ready-for-review". Thanks.

File: CHANGELOG.md [hunk 1]
```diff
@@ -16,6 +16,7 @@ We refer to [GitHub issues](https://github.com/JabRef/jabref/issues) by using `#
 - [#1496](https://github.com/JabRef/jabref/issues/1496) Keep track of which entry a downloaded file belongs to
 - Made it possible to download multiple entries in one action
 - [#1813](https://github.com/JabRef/jabref/issues/1813) Import/Export preferences dialog default directory set to working directory
+- [#1897](https://github.com/JabRef/jabref/issues/1897) Added last four nonpunctuation characters should be numerals in Check Integrity year
```
Comment (Siedlerchr) [inline CHANGELOG.md]: I would reword it a bit to make it a a bit more clear: 
>  Implemented integrity check for year field: Last four nonpunctuation characters should be numerals 

(Put the word "year" in backticks (the french accent grave), just look at other examples in the changelog)

File: src/main/java/net/sf/jabref/logic/integrity/IntegrityCheck.java [hunk 2]
```diff
@@ -306,9 +306,11 @@ private BracketChecker(String field) {
     private static class YearChecker implements Checker {
 
         private static final Predicate<String> CONTAINS_FOUR_DIGIT = Pattern.compile("([^0-9]|^)[0-9]{4}([^0-9]|$)").asPredicate();
+        private static final Predicate<String> ENDS_WITH_FOUR_DIGIT = Pattern.compile("[0-9][0-9][0-9][0-9]$")
```
Comment (Siedlerchr) [inline src/main/java/net/sf/jabref/logic/integrity/IntegrityCheck.java]: You could use` [0-9]{4}$` but it's essentially the same.

**Comment by Siedlerchr** [comment]
Thanks for your work :+1: LGTM! 
If you reword the changelog, we can merge it in.

File: src/main/java/net/sf/jabref/logic/integrity/IntegrityCheck.java [hunk 3]
```diff
@@ -321,6 +323,12 @@ private BracketChecker(String field) {
                 return Collections.singletonList(new IntegrityMessage(Localization.lang("should contain a four digit number"), entry, FieldName.YEAR));
             }
```
Comment (stefan-kolb) [inline src/main/java/net/sf/jabref/logic/integrity/IntegrityCheck.java]: We should add a comment to the BinteX doc here, because this decision on year structure is not intuitive.

File: src/main/java/net/sf/jabref/logic/integrity/IntegrityCheck.java [hunk 4]
```diff
@@ -321,6 +323,12 @@ private BracketChecker(String field) {
...
 
+            if (!ENDS_WITH_FOUR_DIGIT.test(value.get().replaceAll("[](){},.;!?<>%&§]", ""))) {
```
Comment (stefan-kolb) [inline src/main/java/net/sf/jabref/logic/integrity/IntegrityCheck.java]: Where comes the decision on these chars?

Comment (stefan-kolb) [inline src/main/java/net/sf/jabref/logic/integrity/IntegrityCheck.java]: Please define them as punctuation chars as a final variable.

**Comment by tobiasdiez** [comment]
Just a probably stupid question: why not just check that the year field consists exactly of 4 digits?

File: src/main/java/net/sf/jabref/logic/integrity/IntegrityCheck.java (hunk 4 above)
Comment (grimes2) [inline src/main/java/net/sf/jabref/logic/integrity/IntegrityCheck.java]: Some regex chars have a special function, they are ommited. `String.replaceAll(String regex, String replacement)` I failed to define a variable.

**Comment by Siedlerchr** [comment]
@tobiasdiez  Because the offical bibtex spec says so:
See the issue for details:
https://github.com/JabRef/jabref/issues/1897#issuecomment-243756785

File: src/main/java/net/sf/jabref/logic/integrity/IntegrityCheck.java (hunk 4 above)
Comment (Siedlerchr) [inline src/main/java/net/sf/jabref/logic/integrity/IntegrityCheck.java]: And I tink you mixed one up  must be `$ instead of §`

File: src/main/java/net/sf/jabref/logic/integrity/IntegrityCheck.java [hunk 5]
```diff
@@ -306,8 +306,9 @@ private BracketChecker(String field) {
     private static class YearChecker implements Checker {
 
         private static final Predicate<String> CONTAINS_FOUR_DIGIT = Pattern.compile("([^0-9]|^)[0-9]{4}([^0-9]|$)").asPredicate();
-        private static final Predicate<String> ENDS_WITH_FOUR_DIGIT = Pattern.compile("[0-9][0-9][0-9][0-9]$")
+        private static final Predicate<String> ENDS_WITH_FOUR_DIGIT = Pattern.compile("[0-9]{4}$")
                 .asPredicate();
+        private static final String regex = "[](){},.;!?<>%&$]";
```
Comment (stefan-kolb) [inline src/main/java/net/sf/jabref/logic/integrity/IntegrityCheck.java]: Please uppercase constant name and call it something like `PUNCTUATION_MARKS`. Just calling it regex doesn't really help in understanding what it's for.

File: src/main/java/net/sf/jabref/logic/integrity/IntegrityCheck.java [hunk 6]
```diff
@@ -306,9 +306,12 @@ private BracketChecker(String field) {
     private static class YearChecker implements Checker {
 
         private static final Predicate<String> CONTAINS_FOUR_DIGIT = Pattern.compile("([^0-9]|^)[0-9]{4}([^0-9]|$)").asPredicate();
+        private static final Predicate<String> ENDS_WITH_FOUR_DIGIT = Pattern.compile("[0-9]{4}$")
+                .asPredicate();
+        private static final String regex = "[](){},.;!?<>%&$]";
 
         /**
-         * Checks, if the number String contains a four digit year
+         * Checks, if the number String contains a four digit year and ends with it
```
Comment (stefan-kolb) [inline src/main/java/net/sf/jabref/logic/integrity/IntegrityCheck.java]: Please add a comment pointing to the BiBTeX manual where this decision to accept around 1997 and so on is explained.

File: src/main/java/net/sf/jabref/logic/integrity/IntegrityCheck.java [hunk 7]
```diff
@@ -321,6 +324,12 @@ private BracketChecker(String field) {
                 return Collections.singletonList(new IntegrityMessage(Localization.lang("should contain a four digit number"), entry, FieldName.YEAR));
             }
 
+            if (!ENDS_WITH_FOUR_DIGIT.test(value.get().replaceAll(regex, ""))) {
+                return Collections.singletonList(new IntegrityMessage(
+                        Localization.lang("last four nonpunctuation characters should be numerals"), entry,
```
Comment (stefan-kolb) [inline src/main/java/net/sf/jabref/logic/integrity/IntegrityCheck.java]: Localization Message should start with Uppercase character?!

**Comment by stefan-kolb** [comment]
Got 3 minor comments and then we can merge :+1:

File: src/test/java/net/sf/jabref/logic/integrity/IntegrityCheckTest.java [hunk 8]
```diff
@@ -45,9 +45,13 @@ public void testUrlChecks() {
     public void testYearChecks() {
         assertCorrect(createContext("year", "2014"));
         assertCorrect(createContext("year", "1986"));
+        assertCorrect(createContext("year", "around 1986"));
+        assertCorrect(createContext("year", "(around 1986)"));
         assertWrong(createContext("year", "abc"));
         assertWrong(createContext("year", "86"));
         assertWrong(createContext("year", "204"));
+        assertWrong(createContext("year", "1986a"));
```
Comment (Siedlerchr) [inline src/test/java/net/sf/jabref/logic/integrity/IntegrityCheckTest.java]: Please add also a test case with one of the regex symbols

File: src/main/java/net/sf/jabref/logic/integrity/IntegrityCheck.java (hunk 7 above)
Comment (grimes2) [inline src/main/java/net/sf/jabref/logic/integrity/IntegrityCheck.java]: Most of the Check Integrity messages start with lower-case. It is an inconsistency.

File: src/main/java/net/sf/jabref/logic/integrity/IntegrityCheck.java [hunk 9]
```diff
@@ -321,6 +324,12 @@ private BracketChecker(String field) {
                 return Collections.singletonList(new IntegrityMessage(Localization.lang("should contain a four digit number"), entry, FieldName.YEAR));
             }
 
+            if (!ENDS_WITH_FOUR_DIGIT.test(value.get().replaceAll(regex, ""))) {
```
Comment (Siedlerchr) [inline src/main/java/net/sf/jabref/logic/integrity/IntegrityCheck.java]: Something bothered me with the replaceall and the list of regex chars...
After thinking about it for a while I found why:
The first argument of replaceAll is a regex, so your regex char string would be interpreted as actual pattern.
To prevent this, you have to escape the chars, or better use http://docs.oracle.com/javase/8/docs/api/java/util/regex/Pattern.html#quote-java.lang.String-

Comment (grimes2) [inline src/main/java/net/sf/jabref/logic/integrity/IntegrityCheck.java]: The pattern [abc] is needed: Set definition, can match the letter a or b or c.

File: src/main/java/net/sf/jabref/logic/integrity/IntegrityCheck.java (hunk 7 above)
Comment (oscargus) [inline src/main/java/net/sf/jabref/logic/integrity/IntegrityCheck.java]: Most messages do actually start with a lower case character right now so
for consistency it is currently correct. Although I think that conceptually
it should be upper case.

**Comment by stefan-kolb** [comment]
Thanks :smile:  LGTM :+1:

**Comment by Siedlerchr** [comment]
LGTM, too.
